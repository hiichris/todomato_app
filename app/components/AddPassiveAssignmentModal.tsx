import React, { useEffect, useState, useRef } from "react";
import {
  Modal,
  View,
  Text,
  Alert,
  Pressable,
  StyleSheet,
  Dimensions,
  TextInput,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
  TouchableWithoutFeedback,
  Keyboard,
} from "react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { v4 as uuidv4 } from "uuid";
import { Link } from "expo-router";
import { useSQLiteContext } from "expo-sqlite";
import { SafeAreaView } from "react-native-safe-area-context";
import { primaryColor } from "../helpers/constants";
import {
  addNewTask,
  addNewTodo,
  getCategories,
  getTodoById,
} from "../services/db_service";
import { SQLiteProvider } from "expo-sqlite";
import { Todo } from "../models/todo";
import { sendPassiveAssignmentRequest } from "../services/api_service";

export default function AddPassiveAssignmentModal({
  modalVisible,
  setModalVisible,
  setTodos,
  refreshTodos,
  router,
  categories,
  setCategories,
}) {
  const [todoTitle, setTodoTitle] = React.useState("");
  const [selectedCategory, setSelectedCategory] = useState(null);
  const [yourName, setYourName] = React.useState("");
  const [asigneeName, setAsigneeName] = React.useState("");
  const [taskDescription, setTaskDescription] = React.useState("");
  const [asigneeEmail, setAsigneeEmail] = React.useState("");
  const scrollViewRef = useRef(null);
  const inputRef = useRef(null);
  const [uid, setUid] = useState("");

  useEffect(() => {
    const keyboardDidShowListener = Keyboard.addListener(
      "keyboardDidShow",
      () => {
        inputRef.current?.measureLayout(scrollViewRef.current, (x, y) => {
          scrollViewRef.current?.scrollTo({
            y: Dimensions.get("window").height - y - 20,
            animated: true,
          });
        });
      }
    );

    return () => {
      keyboardDidShowListener.remove();
    };
  }, []);

  const generateRandomId = (length = 16) => {
    const characters =
      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789.!^*_-";
    let result = "";
    const charactersLength = characters.length;

    for (let i = 0; i < length; i++) {
      const randomIndex = Math.floor(Math.random() * charactersLength);
      result += characters.charAt(randomIndex);
    }

    return result;
  };

  const cleanupAndCloseModal = () => {
    setTodoTitle("");
    setSelectedCategory(null);
    setYourName("");
    setAsigneeName("");
    setTaskDescription("");
    setAsigneeEmail("");
    setModalVisible(!modalVisible);
  };

  const createPassiveAssignmentHandler = async () => {
    if (todoTitle === "") {
      Alert.alert("Todo Title is required");
      return;
    }
    if (selectedCategory === null) {
      Alert.alert("Please select a category");
      return;
    }
    if (yourName === "") {
      Alert.alert("Your Name is required");
      return;
    }
    if (asigneeName === "") {
      Alert.alert("Asignee Name is required");
      return;
    }
    if (taskDescription === "") {
      Alert.alert("Task Description is required");
      return;
    }
    const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    if (asigneeEmail === "" || !emailRegex.test(asigneeEmail)) {
      Alert.alert("Asignee Email is required and must be a valid email.");
      return;
    }

    console.log(
      "Creating new todo: ",
      todoTitle,
      selectedCategory,
      yourName,
      asigneeName,
      taskDescription,
      asigneeEmail
    );

    addNewTodo(setTodos, todoTitle, selectedCategory)
      .then((result) => {
        console.log("result: ", result);
        refreshTodos();

        // Get last inserted todo_id
        const todo_id = result?.lastInsertRowId;

        // Send the task to Passive Assignment Service API
        const uid = generateRandomId();

        // Add new task
        addNewTask(
          () => {},
          todo_id,
          taskDescription,
          0,
          new Date().toISOString(),
          1,
          0,
          uid
        )
          .then((result) => {
            console.log("Task added: ", result);
            const task_id = result?.lastInsertRowId;

            // Get todo details by id
            getTodoById(todo_id)
              .then((todo: Todo) => {
                console.log("Todo details: ", todo);

                // Clear the todoTitle
                cleanupAndCloseModal();

                sendPassiveAssignmentRequest(
                  uid,
                  todo_id,
                  task_id,
                  taskDescription,
                  yourName,
                  asigneeName,
                  asigneeEmail
                )
                  .then((response) => {
                    if (response) {
                      console.log(
                        "Passive Assignment Request Sent: ",
                        response
                      );
                    }
                  })
                  .catch((error) => {
                    console.log(
                      "Error sending passive assignment request: ",
                      error
                    );
                  });

                // Redirect to todo details page
                router.push({
                  pathname: "/todo_details",
                  params: {
                    title: todo.title,
                    id: todo.id,
                    todoNotes: todo.notes,
                    categoryName: todo.category_name,
                    categoryColor: todo.category_color,
                    has_completed: todo.has_completed,
                    refreshTodos: refreshTodos,
                  },
                });
              })
              .catch((error) => {
                console.log("Error during getTodoById: ", error);
              });
          })
          .catch((error) => {
            console.log("Error during addNewTask: ", error);
          });
      })
      .catch((error) => {
        console.log("Error during addNewTodo: ", error);
      });

    // Clear the todoTitle
    cleanupAndCloseModal();
  };

  const retrieveDBCategories = async () => {
    console.log("Retrieving categories");
    const dbCategories = await getCategories(false);
    setCategories(dbCategories);
  };

  const onFocusHandler = () => {
    if (selectedCategory === null) {
      Alert.alert("Please select a category first!");
      Keyboard.dismiss();
      return;
    }
  };

  return (
    <SafeAreaView>
      <Modal
        animationType="slide"
        transparent={true}
        visible={modalVisible}
        onRequestClose={() => {
          Alert.alert("Modal has been closed.");
          setModalVisible(!modalVisible);
        }}
      >
        <KeyboardAvoidingView
          behavior={Platform.OS === "ios" ? "padding" : "height"}
          style={styles.centeredView}
        >
          <ScrollView
            style={styles.centeredView}
            ref={scrollViewRef}
            keyboardShouldPersistTaps="always"
          >
            <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
              <View style={styles.modalView}>
                <View style={styles.modalHeaderContainer}>
                  <Text style={styles.headerText}>Passive Assignment</Text>
                  <View style={styles.circleButtonContainer}>
                    <Pressable
                      style={styles.closeCirleButton}
                      onPress={() => setModalVisible(!modalVisible)}
                    >
                      <Text style={styles.closeCircleText}>X</Text>
                    </Pressable>
                  </View>
                </View>
                <View style={styles.modalBodyContainer}>
                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Category:</Text>
                    <View style={styles.inputGroupContainer}>
                      <ScrollView
                        style={styles.categoryScrollView}
                        horizontal={true}
                        showsHorizontalScrollIndicator={false}
                      >
                        <View style={styles.categoryContainer}>
                          {categories &&
                            categories.map((category) => (
                              <Pressable
                                key={category.id}
                                style={({ pressed }) => [
                                  styles.categoryButton,
                                  {
                                    backgroundColor: pressed
                                      ? "lightgray"
                                      : "white",
                                  },
                                  {
                                    borderColor:
                                      selectedCategory === category.id
                                        ? "#B2361B"
                                        : "white",
                                  },
                                ]}
                                onPress={() => setSelectedCategory(category.id)}
                              >
                                <Text style={styles.categoryButtonText}>
                                  {category.name}
                                </Text>
                              </Pressable>
                            ))}
                        </View>
                      </ScrollView>
                    </View>
                    <View style={styles.spacer}></View>
                  </View>

                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Todo Title:</Text>
                    <TextInput
                      style={styles.modalTextInput}
                      onChangeText={setTodoTitle}
                      onFocus={onFocusHandler}
                      ref={inputRef}
                      value={todoTitle}
                      textAlign="left"
                      textAlignVertical="top"
                    />
                  </View>

                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>
                      Your Name / Asignee Name:
                    </Text>
                    <View style={styles.inputGroupContainer}>
                      <TextInput
                        style={styles.modalTextInputRow}
                        onChangeText={setYourName}
                        onFocus={onFocusHandler}
                        ref={inputRef}
                        value={yourName}
                        numberOfLines={1}
                        textAlign="left"
                        textAlignVertical="top"
                      />
                      <TextInput
                        style={styles.modalTextInputRow}
                        onChangeText={setAsigneeName}
                        onFocus={onFocusHandler}
                        ref={inputRef}
                        value={asigneeName}
                        numberOfLines={1}
                        textAlign="left"
                        textAlignVertical="top"
                      />
                    </View>
                  </View>

                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Task Description:</Text>
                    <TextInput
                      style={styles.modalTextInput}
                      onChangeText={setTaskDescription}
                      onFocus={onFocusHandler}
                      ref={inputRef}
                      value={taskDescription}
                      numberOfLines={1}
                      textAlign="left"
                      textAlignVertical="top"
                    />
                  </View>

                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Asignee Email:</Text>
                    <TextInput
                      style={styles.modalTextInput}
                      onChangeText={setAsigneeEmail}
                      onFocus={onFocusHandler}
                      ref={inputRef}
                      value={asigneeEmail}
                      numberOfLines={1}
                      textAlign="left"
                      textAlignVertical="top"
                    />
                  </View>

                  <View style={styles.buttonsContainer}>
                    <Pressable
                      style={[styles.button, styles.buttonCreate]}
                      onPress={createPassiveAssignmentHandler}
                    >
                      <Text style={styles.createButtonText}>Create</Text>
                    </Pressable>
                  </View>
                </View>

                <View style={{ flexGrow: 2 }}></View>
              </View>
            </TouchableWithoutFeedback>
          </ScrollView>
        </KeyboardAvoidingView>
      </Modal>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
  },
  modalView: {
    flex: 1,
    width: Dimensions.get("window").width,
    height: Dimensions.get("window").height * 0.75,
    marginTop: Dimensions.get("window").height / 4,
    backgroundColor: "rgba(255, 87, 51, 0.95)",
    borderRadius: 20,
    padding: 16,
    paddingTop: 40,
    alignItems: "center",
    shadowColor: "#000",
    shadowOpacity: 0.15,
    shadowRadius: 5,
    elevation: 5,
    justifyContent: "space-evenly",
  },
  buttonsContainer: {
    margin: 8,
    flexDirection: "row",
    justifyContent: "space-between",
    marginTop: 8,
  },
  button: {
    borderRadius: 20,
    padding: 16,
    paddingHorizontal: 12,
    margin: 8,
    elevation: 2,
  },
  buttonCreate: {
    backgroundColor: "#B2361B",
    borderColor: primaryColor,
    width: "100%",
  },
  createButtonText: {
    color: "white",
    fontWeight: "bold",
    textAlign: "center",
    fontSize: 14,
  },
  modalText: {
    flex: 3,
    textAlign: "center",
    color: "white",
    fontSize: 14,
    fontWeight: "bold",
  },
  modalTextDescription: {
    flex: 2,
    fontSize: 8,
    color: "white",
    paddingTop: 8,
  },
  modalTextInput: {
    flex: 4,
    borderColor: "white",
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    margin: 8,
    color: "white",
    height: 40,
    verticalAlign: "top",
    justifyContent: "flex-start",
  },
  modalTextInputRow: {
    flex: 4,
    borderColor: "white",
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    margin: 10,
    color: "white",
    height: 40,
    justifyContent: "flex-start",
  },
  inputContiner: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  modalHeaderContainer: {
    flex: 2,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  modalBodyContainer: {
    flex: 50,
    justifyContent: "flex-start",
  },
  headerText: {
    color: "white",
    fontSize: 18,
    fontWeight: "bold",
    marginTop: -38,
  },
  circleButtonContainer: {
    top: -20,
    position: "absolute",
    width: "100%",
    justifyContent: "center",
    alignItems: "center",
  },
  closeCirleButton: {
    left: Dimensions.get("window").width / 2 - 30,
    backgroundColor: "white",
    borderRadius: 20,
    width: 30,
    height: 30,
    justifyContent: "center",
    alignItems: "center",
  },
  closeCircleText: {
    color: primaryColor,
    fontWeight: "bold",
    fontSize: 10,
    width: 13,
    textAlign: "center",
  },
  inputGroupContainer: {
    flex: 5,
    flexDirection: "row",
    justifyContent: "center",
    alignContent: "center",
  },
  categoryContainer: {
    padding: 0,
    flexDirection: "row",
    height: 50,
    width: "100%",
  },
  categoryButton: {
    flex: 1,
    margin: 4,
    backgroundColor: "white",
    padding: 12,
    paddingVertical: 0,
    borderRadius: 20,
    justifyContent: "center",
    borderWidth: 4,
  },
  categoryButtonText: {
    color: primaryColor,
    fontSize: 14,
    textAlign: "center",
  },
  categoryScrollView: {
    marginRight: 8,
  },
  spacer: {
    height: 100,
  },
  passAssignmentContainer: {
    flex: 1,
    flexDirection: "column",
    textAlign: "center",
    marginHorizontal: 16,
    marginTop: 16,
  },
});
