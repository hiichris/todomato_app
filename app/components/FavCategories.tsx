import {
  View,
  Text,
  StyleSheet,
  Dimensions,
  ScrollView,
  Pressable,
} from "react-native";
import { getCategories } from "../services/db_service";
import { useState, useEffect } from "react";

export const FavCategories = ({setSearchQuery}) => {
  const [categories, setCategories] = useState([]);

  const retrieveDBCategories = async () => {
    console.log("Retrieving categories");
    const dbCategories = await getCategories(true);
    console.log("dbCategories: ", dbCategories);
    setCategories(dbCategories);
  };

  const screenWidth = Dimensions.get("window").width; // Get the screen width

  useEffect(() => {
    retrieveDBCategories();
  }, []);

  return (
    <View style={styles.favCategoryContainer}>
      <Text style={styles.favCategoryText}>Fav Categories</Text>

      <ScrollView
        style={styles.favButtonsContainer}
        horizontal={true}
        showsHorizontalScrollIndicator={false}
      >
        {categories &&
          categories.map((category) => (
            <Pressable
              key={category.id}
              style={[
                styles.favButton,
                {
                  backgroundColor: category.color,
                },
              ]}
              onPress={() => {
                console.log("Category pressed: ", category.id);
                setSearchQuery(category.name);
              }}
            >
              <Text style={styles.favButtonText}>{category.name}</Text>
            </Pressable>
          ))}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  favCategoryContainer: {
    flex: 1,
    marginHorizontal: 18,
  },
  favCategoryText: {
    fontSize: 16,
    fontWeight: "bold",
    color: "black",
    marginVertical: 8,
  },
  favButtonsContainer: {
    flex: 1,
    flexDirection: "row",
    padding: 0,
    margin: 0,
  },
  favButton: {
    borderRadius: 50,
    margin: 2,
    width: 100,
    justifyContent: "center",
    alignItems: "center",
  },
  favButtonText: {
    color: "white",
    fontSize: 10,
    fontWeight: "bold",
  },
});
