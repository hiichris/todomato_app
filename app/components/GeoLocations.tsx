import React from "react";
import {
  View,
  Text,
  TextInput,
  Linking,
  Pressable,
  Platform,
} from "react-native";
import { useState, useRef } from "react";
import MapView, { Marker } from "react-native-maps";
import * as Location from "expo-location";

export const GeoLocations = ({ styles, setFocusedInput }) => {
  const [marker, setMarker] = useState(null);
  const [address, setAddress] = useState("");
  const mapRef = useRef(null);
  const [region, setRegion] = useState({
    latitude: 37.78825,
    longitude: -122.4324,
    latitudeDelta: 0.0922,
    longitudeDelta: 0.0421,
  });

  const findCoordinates = async () => {
    try {
      let geocode = await Location.geocodeAsync(address);
      if (geocode.length > 0) {
        const { latitude, longitude } = geocode[0];
        const newRegion = {
          latitude: latitude,
          longitude: longitude,
          latitudeDelta: 0.0922,
          longitudeDelta: 0.0421,
        };
        mapRef.current?.animateToRegion(newRegion, 1000);
        setRegion(newRegion);
        setMarker({ latitude: latitude, longitude: longitude });
        console.log("Geocode:", { latitude, longitude });
      }
    } catch (error) {
      console.error("Error getting geolocation:", error);
    }
  };

  return (
    <>
      <View style={styles.sectionContainer}>
        <Text style={styles.detailsSectionTitle}>Something to pin</Text>
      </View>

      <View style={styles.geolocContainer}>
        <View style={styles.geolocSerachContainer}>
          <TextInput
            onFocus={() => setFocusedInput("map")}
            placeholder="Enter a location or address..."
            placeholderTextColor={"gray"}
            onChangeText={(text) => setAddress(text)}
            value={address}
            style={styles.geolocInput}
          />
          <Pressable onPress={findCoordinates} style={styles.geolocButton}>
            <Text style={styles.geolocButtonText}>Find Location</Text>
          </Pressable>
        </View>

        <MapView
          region={region}
          onRegionChangeComplete={(region) => setRegion(region)}
          mapType="standard"
          zoomControlEnabled={true}
          ref={mapRef}
          style={styles.geolocMap}
        >
          {marker && (
            <Marker
              coordinate={marker}
              title="📍 Selected Location"
              description={address}
            />
          )}
        </MapView>

        {marker && (
          <Pressable
            style={styles.geolocLinkContainer}
            onPress={() => {
              const latitude = region.latitude;
              const longitude = region.longitude;
              const mapUrl = Platform.select({
                ios: `http://maps.apple.com/?q=${latitude},${longitude}`,
                android: `geo:${latitude},${longitude}?q=${latitude},${longitude}`,
              });
              console.log("mapUrl:", mapUrl);
              Linking.openURL(mapUrl);
            }}
          >
            <Text style={styles.geolocLinkText}>Open in Maps</Text>
          </Pressable>
        )}
      </View>
    </>
  );
};
