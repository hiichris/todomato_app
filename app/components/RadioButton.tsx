import { Text, TouchableOpacity, View, StyleSheet } from "react-native";

// Radio Button Component
export const RadioButton = ({ selected, onPress, label }) => {
  return (
    <TouchableOpacity onPress={onPress} style={styles.radioButtonContainer}>
      <View
        style={[
          styles.radioButton,
          { borderColor: selected ? "white" : "white" },
        ]}
      >
        {selected ? <View style={styles.radioButtonInner} /> : null}
      </View>
      <Text style={styles.labelText}>{label}</Text>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  radioButtonContainer: {
    flexDirection: "row",
    alignItems: "center",
  },
  radioButton: {
    height: 30,
    width: 30,
    borderRadius: 20,
    borderWidth: 2,
    alignItems: "center",
    justifyContent: "center",
    marginRight: 10,
  },
  radioButtonInner: {
    height: 10,
    width: 10,
    borderRadius: 5,
    backgroundColor: "yellow",
  },
  labelText: {
    color: "white",
  },
});
