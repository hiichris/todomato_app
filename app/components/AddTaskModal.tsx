import React, { useEffect, useState, useRef } from "react";
import {
  Modal,
  View,
  Text,
  Alert,
  Pressable,
  StyleSheet,
  Dimensions,
  TextInput,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
  TouchableWithoutFeedback,
  Keyboard,
} from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import { primaryColor } from "../helpers/constants";
import { addNewTask } from "../services/db_service";
import { ToDoDateTimePicker } from "./TodoDateTimePicker";
import { RadioButton } from "./RadioButton";
import moment from "moment";

export default function AddTaskModal({
  modalVisible,
  todoId,
  setModalVisible,
  setTasks,
  refreshTasks,
  scheduleNotification,
}) {
  const [taskName, setTaskName] = React.useState("");
  const [duration, setDuration] = React.useState(5);
  const scrollViewRef = useRef(null);
  const durationInterval = [1, 5, 15, 20, 30, 60];
  const [selectedStartOption, setSelectedStartOption] = useState("now");
  const [selectedDate, setSelectedDate] = useState("");

  useEffect(() => {
    // Add a listener to the keyboard
    const keyboardDidShowListener = Keyboard.addListener(
      "keyboardDidShow",
      () => {
        // Scroll to the bottom of the ScrollView when the keyboard is shown
        scrollViewRef.current?.scrollToEnd({ animated: true });
      }
    );

    return () => {
      keyboardDidShowListener.remove();
    };
  }, []);

  // Get the time difference in seconds
  const getTimeDifferenceInSeconds = (
    dateString: string,
    inputFormat: string = "MMM D YYYY, h:mm A"
  ) => {
    // If the date is empty, set it to today's date
    if (dateString === "" || dateString === null) {
      dateString = moment(new Date()).format(inputFormat);
    }
    // Convert the selected date with today's date and convert it in seconds
    const currentDate = new Date();
    const selectedDateObject = moment(dateString, inputFormat).toDate();
    const differenceInMilliseconds =
      new Date(selectedDateObject).getTime() - currentDate.getTime();
    const differenceInSeconds = Math.floor(differenceInMilliseconds / 1000);

    return [differenceInSeconds, dateString];
  };

  // Add a new task
  const addNewTaskHandler = async () => {
    // Validate the input
    if (taskName === "") {
      Alert.alert("Task name cannot be empty!");
      return;
    }

    if (duration < 1) {
      Alert.alert("Duration must be greater than 0!");
      return;
    }
    // Get the difference in seconds from selected date
    const [differenceInSeconds, selectedDateString] =
      getTimeDifferenceInSeconds(selectedDate);
    console.log("differenceInSeconds: ", differenceInSeconds);

    // If the user selected "schedule" option, we will have to schedule the
    // notification based on the selected date and time. Then, schedule another
    // notification after the duration ends.
    if (selectedStartOption === "scheduled") {
      console.log("seelctedstartoption ", selectedStartOption);
      // Schedule the 1st reminder notification
      scheduleNotification(taskName, differenceInSeconds, "Reminder")
        .then((result) => {
          console.log(
            "Reminder notification scheduled in " +
              differenceInSeconds +
              " seconds!",
            result
          );
        })
        .catch((error) => {
          console.log("Error scheduling reminder notification: ", error);
        });
    }
    // Calculate the total duration in seconds for the alarm notification
    // If the difference is negative, it means there's no selected date and time.
    // We'll use the duration as the total duration for the alarm notification.
    var alaramDuration = 0;
    if (differenceInSeconds < 0) {
      alaramDuration = duration * 60;
    } else {
      alaramDuration = differenceInSeconds + duration * 60;
    }

    console.log("alaramDuration::: ", alaramDuration);

    // Schedule the alarm (2nd) notification
    scheduleNotification(taskName, alaramDuration, "Times up!")
      .then((result) => {
        console.log(
          "Alarm notification scheduled in " + alaramDuration + " seconds!",
          result
        );
        Alert.alert("Task scheduled successfully!ðŸŽ‰");
      })
      .catch((error) => {
        console.log("Error scheduling alram notification: ", error);
      });

    // Add the task to the database
    addNewTask(setTasks, todoId, taskName, duration, selectedDateString)
      .then((result) => {
        console.log("result: ", result);
        refreshTasks();
      })
      .catch((error) => {
        console.log("addNewTask Error: ", error);
      });

    // Clear the fields back to default
    setTaskName("");
    setDuration(5);

    // Close the modal
    setModalVisible(!modalVisible);
  };

  return (
    <SafeAreaView>
      <Modal
        animationType="slide"
        transparent={true}
        visible={modalVisible}
        onRequestClose={() => {
          Alert.alert("Modal has been closed.");
          setModalVisible(!modalVisible);
        }}
      >
        <KeyboardAvoidingView
          behavior={Platform.OS === "ios" ? "padding" : "height"}
          style={styles.centeredView}
        >
          <ScrollView
            style={styles.centeredView}
            ref={scrollViewRef}
            keyboardShouldPersistTaps="always"
          >
            <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
              <View
                style={[
                  styles.modalView,
                  {
                    width: Dimensions.get("window").width,
                    height: Dimensions.get("window").height / 2,
                    marginTop: Dimensions.get("window").height / 2,
                  },
                ]}
              >
                <View style={styles.modalHeaderContainer}>
                  <Text style={styles.headerText}>Create Task</Text>
                  <Pressable
                    style={styles.closeCirleButton}
                    onPress={() => setModalVisible(!modalVisible)}
                  >
                    <Text style={styles.closeCircleText}>X</Text>
                  </Pressable>
                </View>
                <View style={styles.modalBodyContainer}>
                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Task Name:</Text>
                    <TextInput
                      style={styles.modalTextInput}
                      onChangeText={setTaskName}
                      value={taskName}
                    />
                    <Text> </Text>
                  </View>

                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Start From:</Text>
                    <View style={styles.inputGroupContainer}>
                      <RadioButton
                        selected={selectedStartOption === "now"}
                        onPress={() => setSelectedStartOption("now")}
                        label="Now"
                      />
                      <RadioButton
                        selected={selectedStartOption === "scheduled"}
                        onPress={() => setSelectedStartOption("scheduled")}
                        label="Scheduled"
                      />
                    </View>
                    <Text> </Text>
                  </View>

                  {selectedStartOption === "scheduled" && (
                    <View style={styles.inputContiner}>
                      <Text style={styles.modalText}>Notify On:</Text>
                      <ToDoDateTimePicker
                        selectedDate={selectedDate}
                        setSelectedDate={setSelectedDate}
                      />
                    </View>
                  )}

                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Duration (min):</Text>
                    <View style={styles.inputGroupContainer}>
                      {durationInterval.map((number) => (
                        <Pressable
                          key={number}
                          style={({ pressed }) => [
                            styles.durationButton,
                            {
                              backgroundColor: pressed
                                ? "orange"
                                : duration === parseInt(number)
                                ? "#de2500"
                                : "white",
                            },
                          ]}
                          onPress={() => setDuration(parseInt(number))}
                        >
                          <Text
                            key={number}
                            style={[
                              styles.durationText,
                              {
                                color:
                                  duration === parseInt(number)
                                    ? "white"
                                    : primaryColor,
                              },
                            ]}
                          >
                            {number}
                          </Text>
                        </Pressable>
                      ))}
                    </View>
                  </View>

                  <View style={styles.buttonsContainer}>
                    <Pressable
                      style={({ pressed }) => [
                        styles.button,
                        styles.buttonCreate,
                        {
                          backgroundColor: pressed ? "#9a2800" : "#B2361B",
                        },
                      ]}
                      onPress={addNewTaskHandler}
                    >
                      <Text style={styles.createButtonContainer}>Create</Text>
                    </Pressable>
                  </View>
                </View>
                <View style={{ flexGrow: 2 }}></View>
              </View>
            </TouchableWithoutFeedback>
          </ScrollView>
        </KeyboardAvoidingView>
      </Modal>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
  },
  modalView: {
    flex: 1,
    backgroundColor: "rgba(255, 87, 51, 0.95)",
    borderRadius: 20,
    padding: 16,
    paddingTop: 40,
    alignItems: "center",
    shadowColor: "#000",
    shadowOpacity: 0.15,
    shadowRadius: 5,
    elevation: 5,
    justifyContent: "space-evenly",
  },
  buttonsContainer: {
    margin: 8,
    flexDirection: "row",
    justifyContent: "space-between",
  },
  button: {
    borderRadius: 20,
    padding: 8,
    paddingHorizontal: 12,
    margin: 8,
    elevation: 2,
  },
  buttonCreate: {
    backgroundColor: "#B2361B",
    borderColor: primaryColor,
    borderWidth: 1,
    width: "100%",
  },
  createButtonContainer: {
    color: "white",
    fontWeight: "bold",
    textAlign: "center",
    fontSize: 16,
    padding: 8,
  },
  modalText: {
    flex: 2,
    textAlign: "center",
    color: "white",
    fontSize: 14,
    fontWeight: "bold",
  },
  modalTextInput: {
    flex: 4,
    height: 40,
    fontSize: 14,
    width: Dimensions.get("window").width - 32,
    borderColor: "white",
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    margin: 8,
    color: "white",
  },
  inputContiner: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center",
    textAlign: "left",
  },
  inputGroupContainer: {
    flex: 4,
    flexDirection: "row",
    justifyContent: "space-evenly",
  },
  modalHeaderContainer: {
    flex: 2,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  modalBodyContainer: {
    flex: 50,
    justifyContent: "flex-start",
    alignItems: "center",
  },
  headerText: {
    color: "white",
    fontSize: 18,
    fontWeight: "bold",
    marginTop: -38,
  },
  closeCirleButton: {
    marginTop: -38,
    position: "absolute",
    left: Dimensions.get("window").width / 2,
    top: 16,
    backgroundColor: "white",
    borderRadius: 20,
    width: 30,
    height: 30,
    justifyContent: "center",
    alignItems: "center",
  },
  closeCircleText: {
    color: primaryColor,
    fontWeight: "bold",
    fontSize: 10,
    width: 13,
    textAlign: "center",
  },
  durationButton: {
    borderRadius: 50,
    width: 30,
    height: 30,
    textAlign: "center",
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "white",
  },
  durationText: {
    color: primaryColor,
    fontSize: 12,
    fontWeight: "bold",
  },
  instructionContainer: {
    fontSize: 10,
    textAlign: "center",
  },
  instructionText: {
    color: "white",
    marginTop: 8,
    textAlign: "center",
  },
});
