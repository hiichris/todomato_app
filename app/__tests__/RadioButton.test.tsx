import React from "react";
import renderer from "react-test-renderer";
import { waitFor } from "@testing-library/react-native";

import { RadioButton } from "../components/RadioButton";

describe("RadioButton", () => {
  it("renders correctly", async () => {
    const tree = renderer.create(<RadioButton />).toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });
});
