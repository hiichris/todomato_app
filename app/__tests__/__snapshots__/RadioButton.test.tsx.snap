// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 20,
          "borderWidth": 2,
          "height": 30,
          "justifyContent": "center",
          "marginRight": 10,
          "width": 30,
        },
        {
          "borderColor": "white",
        },
      ]
    }
  />
  <Text
    style={
      {
        "color": "white",
      }
    }
  />
</View>
`;
